### PR TITLE
Allows :dql_parameters with ARRAY_* functions

### PR DESCRIPTION
**Pull request by allan-simon**
before it was only possible to do `ARRAY_APPEND(e.myarray, 'a_literal')` so it was impossible to have the value coming from php (without resorting to DQL injection)

We now allow also the following syntax `ARRAY_APPEND(e.myarray, :foobar)`

### CODE REVIEW

**Comment by allan-simon** [comment]
compare to the other PR I've added additional test and used ArithmeticPrimary that allow both Literal and input parameters , as well as other methods returning string (which allow chaining calls )

File: tests/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/ArrayAppendTest.php [hunk 1]
```diff
@@ -21,6 +21,8 @@ protected function getExpectedSqlStatements(): array
         return [
             'SELECT array_append(c0_.array1, 1989) AS sclr_0 FROM ContainsArrays c0_',
             "SELECT array_append(c0_.array1, 'country') AS sclr_0 FROM ContainsArrays c0_",
+            "SELECT array_append(c0_.array1, ?) AS sclr_0 FROM ContainsArrays c0_",
+            "SELECT array_append(c0_.array1, 2 * ?) AS sclr_0 FROM ContainsArrays c0_",
```
Comment (allan-simon) [inline tests/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/ArrayAppendTest.php:25]: sorry this was an intermediate try, i forgot to push my commit --amend , i will do that in a few hours

**Comment by martin-georgiev** [comment]
I think only CI failure remains to be addressed. Thank you 😊

File: tests/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/ArrayPrependTest.php [hunk 2]
```diff
@@ -29,6 +30,7 @@ protected function getDqlStatements(): array
         return [
             \sprintf('SELECT ARRAY_PREPEND(1885, e.array1) FROM %s e', ContainsArrays::class),
             \sprintf("SELECT ARRAY_PREPEND('red', e.array1) FROM %s e", ContainsArrays::class),
+            \sprintf("SELECT ARRAY_PREPEND(:dql_parameter, e.array1) FROM %s e", ContainsArrays::class),
```
Comment (allan-simon) [inline tests/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/ArrayPrependTest.php:33]:
```suggestion
            \sprintf('SELECT ARRAY_PREPEND(:dql_parameter, e.array1) FROM %s e', ContainsArrays::class),
```

File: tests/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/ArrayRemoveTest.php [hunk 3]
```diff
@@ -21,6 +21,7 @@ protected function getExpectedSqlStatements(): array
         return [
             'SELECT array_remove(c0_.array1, 1944) AS sclr_0 FROM ContainsArrays c0_',
             "SELECT array_remove(c0_.array1, 'peach') AS sclr_0 FROM ContainsArrays c0_",
+            "SELECT array_remove(c0_.array1, ?) AS sclr_0 FROM ContainsArrays c0_",
```
Comment (allan-simon) [inline tests/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/ArrayRemoveTest.php:24]:
```suggestion
            'SELECT array_remove(c0_.array1, ?) AS sclr_0 FROM ContainsArrays c0_',
```

File: tests/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/ArrayRemoveTest.php [hunk 4]
```diff
@@ -29,6 +30,7 @@ protected function getDqlStatements(): array
         return [
             \sprintf('SELECT ARRAY_REMOVE(e.array1, 1944) FROM %s e', ContainsArrays::class),
             \sprintf("SELECT ARRAY_REMOVE(e.array1, 'peach') FROM %s e", ContainsArrays::class),
+            \sprintf("SELECT ARRAY_REMOVE(e.array1, :dql_parameter) FROM %s e", ContainsArrays::class),
```
Comment (allan-simon) [inline tests/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/ArrayRemoveTest.php:33]:
```suggestion
            \sprintf('SELECT ARRAY_REMOVE(e.array1, :dql_parameter) FROM %s e', ContainsArrays::class),
```

File: tests/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/ArrayReplaceTest.php [hunk 5]
```diff
@@ -21,6 +21,7 @@ protected function getExpectedSqlStatements(): array
         return [
             'SELECT array_replace(c0_.array1, 1939, 1957) AS sclr_0 FROM ContainsArrays c0_',
             "SELECT array_replace(c0_.array1, 'green', 'mint') AS sclr_0 FROM ContainsArrays c0_",
+            "SELECT array_replace(c0_.array1, 'green', ?) AS sclr_0 FROM ContainsArrays c0_",
```
Comment (allan-simon) [inline tests/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/ArrayReplaceTest.php:24]:
```suggestion
            'SELECT array_replace(c0_.array1, 'green', ?) AS sclr_0 FROM ContainsArrays c0_',
```

File: tests/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/ArrayReplaceTest.php [hunk 6]
```diff
@@ -29,6 +30,7 @@ protected function getDqlStatements(): array
         return [
             \sprintf('SELECT ARRAY_REPLACE(e.array1, 1939, 1957) FROM %s e', ContainsArrays::class),
             \sprintf("SELECT ARRAY_REPLACE(e.array1, 'green', 'mint') FROM %s e", ContainsArrays::class),
+            \sprintf("SELECT ARRAY_REPLACE(e.array1, 'green', :dql_parameter) FROM %s e", ContainsArrays::class),
```
Comment (allan-simon) [inline tests/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/ArrayReplaceTest.php:33]:
```suggestion
            \sprintf('SELECT ARRAY_REPLACE(e.array1, 'green', :dql_parameter) FROM %s e', ContainsArrays::class),
```

**Comment by allan-simon** [comment]
@martin-georgiev  it was some coding style issue, should be good now :)

**Review by martin-georgiev — DISMISSED** [review]

File: tests/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/ArrayReplaceTest.php [hunk 7]
```diff
@@ -21,6 +21,7 @@ protected function getExpectedSqlStatements(): array
         return [
             'SELECT array_replace(c0_.array1, 1939, 1957) AS sclr_0 FROM ContainsArrays c0_',
             "SELECT array_replace(c0_.array1, 'green', 'mint') AS sclr_0 FROM ContainsArrays c0_",
+            'SELECT array_replace(c0_.array1, 'green', ?) AS sclr_0 FROM ContainsArrays c0_',
```
Comment (allan-simon) [inline tests/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/ArrayReplaceTest.php:24]:
```suggestion
            "SELECT array_replace(c0_.array1, 'green', ?) AS sclr_0 FROM ContainsArrays c0_",
```

File: tests/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/ArrayReplaceTest.php [hunk 8]
```diff
@@ -29,6 +30,7 @@ protected function getDqlStatements(): array
         return [
             \sprintf('SELECT ARRAY_REPLACE(e.array1, 1939, 1957) FROM %s e', ContainsArrays::class),
             \sprintf("SELECT ARRAY_REPLACE(e.array1, 'green', 'mint') FROM %s e", ContainsArrays::class),
+            \sprintf('SELECT ARRAY_REPLACE(e.array1, 'green', :dql_parameter) FROM %s e', ContainsArrays::class),
```
Comment (allan-simon) [inline tests/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/ArrayReplaceTest.php:33]:
```suggestion
            \sprintf("SELECT ARRAY_REPLACE(e.array1, 'green', :dql_parameter) FROM %s e", ContainsArrays::class),
```

**Comment by allan-simon** [comment]
@martin-georgiev  sorry, i was a bit fast in the correction of the coding style, it's ok now, the scrutinizer error is a timeout so i think it's just need a relaunch ?

**Review by martin-georgiev — APPROVED** [review]

**Pull request merged**